File: daemon/src/features/agents/Agents.ts
import { AGENT_ROUNDS } from '@daemon/constants/agents'
import path from 'node:path'
import type { ToolSet } from 'ai'
import type { Chat } from '@daemon/types/api/chat'
import { CHAT_MODELS } from '@daemon/types/api/chat'
import type {
  Agent,
  AgentInOrg,
  AgentSummary,
  NewAgent,
} from '@daemon/types/api/agents'
import type { Persona } from '@daemon/types/api/personas'
import type { Tree } from '@daemon/services/Tree'
import { ChatError } from '../chat/error'
import type { Chats, Turn } from '@daemon/features/chat/Chats'
import type { ChatStore } from '../chat/db'
import { agentBrief } from './brief'
import { agentTools, type AgentToolDeps } from './tools'

export interface AgentSite {
  path: string
  tree: Tree
  personas: Persona[]
}

export interface AgentsDeps {
  store: ChatStore
  chats: Chats
  /** The open project, asked each time. Null is no project, which is no agents. */
  project: () => AgentSite | null
  /** The persona's body by name, or null when it is not there. */
  persona: (name: string) => Promise<string | null>
  profile: () => string | null
  /** The app brief for an agent's turn — the room, as the agent surface tells it. */
  brief: () => string
  /** The same room as a terminal is told it, for the Claude Code errands: they have a shell. */
  terminalBrief: () => string
  /** The checkout the hands work in, asked per call — the scoped one, where a shell opens. */
  checkout: () => string
  env: () => Record<string, string>
  tools: Omit<AgentToolDeps, 'checkout' | 'env' | 'brief' | 'persona' | 'name' | 'attachments' | 'progress'>
}

/**
 * Agents: who there is, and how each of them takes a turn.
 *
 * The thread itself is the chat page's — an agent's conversation is a chat row marked as
 * theirs, opened on the same socket, written by the same reply bookkeeping. What this adds is
 * the person: which persona speaks, which hands it has, and where its work goes.
 */
export class Agents {
  constructor(private readonly deps: AgentsDeps) {}

  list(): { agents: AgentSummary[] } {
    const project = this.deps.project()
    if (!project) return { agents: [] }
    return { agents: this.deps.store.agents(project.path).map((one) => this.summarize(one)) }
  }

  /** The chart: everyone, with who they report to and where they stand. */
  org(): { agents: AgentInOrg[] } {
    const project = this.deps.project()
    if (!project) return { agents: [] }
    return { agents: this.deps.store.org(project.path).map((one) => this.summarize(one)) }
  }

  /**
   * Who somebody reports to, or nobody. A loop is refused by walking upward from the
   * proposed lead: meeting the agent on the way is the line closing on itself, and a chart
   * asked who to escalate to would have no answer. Both ends have to be in this project —
   * the chart is per-project the way the agents are.
   */
  setLead(agent: string, lead: string | null): void {
    const project = this.deps.project()
    if (!project) throw new ChatError('no project is open')
    const chart = new Map(this.deps.store.org(project.path).map((one) => [one.id, one]))
    const held = chart.get(agent)
    if (!held) throw new ChatError('no such agent')
    if (lead === null) return this.deps.store.setLead(agent, null)
    if (lead === agent) throw new ChatError(`${held.name} cannot report to themselves`)
    const above = chart.get(lead)
    if (!above) throw new ChatError('no such agent')
    for (let step = above.lead; step; step = chart.get(step)?.lead ?? null)
      if (step === agent)
        throw new ChatError(
          `that would make a loop: ${above.name} already reports to ${held.name}`,
        )
    this.deps.store.setLead(agent, lead)
  }

  /** Where they stand, after a drag. */
  place(agent: string, x: number, y: number): void {
    this.deps.store.place(this.require(agent).id, x, y)
  }

  /**
   * A new colleague. The persona has to be one the project carries and the model one the app
   * serves — an agent made with neither would be a name that answers nothing. Their
   * attachments folder is made now, so it is in the tree from the first message and there is
   * a place to point at before anything is in it.
   */
  async create(input: NewAgent): Promise<Agent> {
    const project = this.deps.project()
    if (!project) throw new ChatError('no project is open')
    const name = input.name.trim()
    if (!name) throw new ChatError('an agent needs a name')
    if (!project.personas.some((one) => one.name === input.persona))
      throw new ChatError(`no persona called ${input.persona} in this project`)
    if (!CHAT_MODELS.some((one) => one.id === input.model))
      throw new ChatError(`no such model: ${input.model}`)
    const made = this.deps.store.createAgent(project.path, { ...input, name })
    await project.tree.mkdir(made.attachments)
    return made
  }

  remove(id: string): void {
    const held = this.require(id)
    this.deps.chats.remove(held.chat)
    this.deps.store.removeAgent(held.id)
  }

  clear(id: string): void {
    this.deps.chats.clear(this.require(id).chat)
  }

  /**
   * Which model answers as them, changed on someone who already exists. Refused for a model
   * the app does not serve, the way hiring is — an agent pointed at nothing is a name that
   * answers nothing, and it would only be found out at the next thing said to them.
   */
  setModel(id: string, model: string): Agent {
    const held = this.require(id)
    if (!CHAT_MODELS.some((one) => one.id === model))
      throw new ChatError(`no such model: ${model}`)
    this.deps.store.setAgentModel(held.id, model)
    return this.require(held.id)
  }

  /** Whose conversation this is, or null for the page's own. */
  of(chat: Chat): Agent | null {
    return this.deps.store.agentOfChat(chat.id)
  }

  /**
   * How an agent answers: the app brief in the agent's room, then who they are and how
   * they talk, with hands that know whose they are. Built each turn, since the persona on
   * disk and the checkout under the hands both move while a conversation stays open.
   */
  async turn(
    agent: Agent,
    progress: (toolCallId: string, note: string) => void,
  ): Promise<Turn> {
    const project = this.deps.project()
    const personaBody = await this.deps.persona(agent.persona)
    const attachmentsAbs = project
      ? path.join(project.path, agent.attachments)
      : agent.attachments
    const system = agentBrief(this.deps.brief(), {
      name: agent.name,
      persona: agent.persona,
      personaBody,
      profile: this.deps.profile(),
      attachmentsAbs,
      attachments: agent.attachments,
    })
    const tools: ToolSet = agentTools({
      ...this.deps.tools,
      checkout: this.deps.checkout,
      env: this.deps.env,
      brief: this.deps.terminalBrief,
      persona: personaBody,
      name: agent.name,
      by: `agent/${agent.name}`,
      attachments: attachmentsAbs,
      progress,
    })
    return { system, tools, maxRounds: AGENT_ROUNDS }
  }

  /** An agent as a page asks for one: what they are, plus whether they are answering right
   *  now and when they last did. */
  private summarize<T extends Agent>(one: T): T & { working: boolean; lastAt: number | null } {
    return {
      ...one,
      working: this.deps.chats.working(one.chat),
      lastAt: this.deps.store.lastSaidAt(one.chat),
    }
  }

  private require(id: string): Agent {
    const held = this.deps.store.agent(id)
    if (!held) throw new ChatError('no such agent')
    return held
  }
}

File: daemon-go/internal/app/agents.go
// The people-shaped agents under the chats: who there is, whose thread is whose, and the chart.
//
// How one answers is `agentturn.go`; what this holds is who they are.

package app

import (
	"strings"

	"github.com/broodmotherai/broodmother/daemon-go/internal/apperr"
	"github.com/broodmotherai/broodmother/daemon-go/internal/chat"
	"github.com/broodmotherai/broodmother/daemon-go/internal/doc"
	"github.com/broodmotherai/broodmother/daemon-go/internal/personas"
)

// AgentSummary is an agent as a list draws one.
type AgentSummary struct {
	chat.Agent
	// Working is whether a reply of theirs is being written right now — the presence dot.
	Working bool `json:"working"`
	// LastAt is when the last thing was said in their thread, or nil when nothing has been.
	LastAt *int64 `json:"lastAt"`
}

// AgentInOrg is the same, with where they stand.
type AgentInOrg struct {
	AgentSummary
	Lead  *string     `json:"lead"`
	Place *chat.Place `json:"place"`
}

// ListAgents is every agent in the open project, by name. Nowhere to work is nobody rather than
// an error, the way the conversations answer.
func (c *Context) ListAgents() []AgentSummary {
	open := c.Workspace.Project()
	if open == nil || c.Chats == nil {
		return []AgentSummary{}
	}
	held := c.Chats.Agents(open.Path)
	found := make([]AgentSummary, 0, len(held))
	for _, one := range held {
		found = append(found, c.summarize(one))
	}
	return found
}

// Org is the chart: everyone, with who they report to and where they stand.
func (c *Context) Org() []AgentInOrg {
	open := c.Workspace.Project()
	if open == nil || c.Chats == nil {
		return []AgentInOrg{}
	}
	held := c.Chats.Org(open.Path)
	found := make([]AgentInOrg, 0, len(held))
	for _, one := range held {
		found = append(found, AgentInOrg{
			AgentSummary: c.summarize(one.Agent), Lead: one.Lead, Place: one.Place,
		})
	}
	return found
}

func (c *Context) summarize(one chat.Agent) AgentSummary {
	working := c.Live != nil && c.Live.Working(one.Chat)
	return AgentSummary{Agent: one, Working: working, LastAt: c.Chats.LastSaidAt(one.Chat)}
}

// AddAgent is a new colleague. The persona has to be one the project carries and the model one
// the app serves — an agent made with neither would be a name that answers nothing. Their
// attachments folder is made now, so it is in the tree from the first message and there is a
// place to point at before anything is in it.
func (c *Context) AddAgent(input chat.NewAgent) (chat.Agent, error) {
	open, err := c.Workspace.RequireProject()
	if err != nil {
		return chat.Agent{}, err
	}
	if c.Chats == nil {
		return chat.Agent{}, apperr.Chatf("there is nowhere to keep an agent")
	}
	input.Name = strings.TrimSpace(input.Name)
	if input.Name == "" {
		return chat.Agent{}, apperr.Chatf("an agent needs a name")
	}
	held, err := c.Root(doc.Project)
	if err != nil {
		return chat.Agent{}, err
	}
	if !carries(held.Personas(), input.Persona) {
		return chat.Agent{}, apperr.Chatf("no persona called %s in this project", input.Persona)
	}
	if chat.ProviderOf(input.Model) == "" {
		return chat.Agent{}, apperr.Chatf("no such model: %s", input.Model)
	}

	made, err := c.Chats.CreateAgent(open.Path, input)
	if err != nil {
		return chat.Agent{}, err
	}
	held.Tree.Mkdir(made.Attachments)
	return made, nil
}

func carries(held []personas.Persona, name string) bool {
	for _, one := range held {
		if one.Name == name {
			return true
		}
	}
	return false
}

// RemoveAgent takes the agent and their conversation together; the attachments folder stays,
// since what they made is yours.
func (c *Context) RemoveAgent(id string) error {
	held, err := c.requireAgent(id)
	if err != nil {
		return err
	}
	if err := c.RemoveChat(held.Chat); err != nil {
		return err
	}
	c.Chats.RemoveAgent(held.ID)
	return nil
}

// ClearAgent takes what was said in their thread; the agent stays to be said to again.
func (c *Context) ClearAgent(id string) error {
	held, err := c.requireAgent(id)
	if err != nil {
		return err
	}
	c.Live.Clear(held.Chat)
	return nil
}

// SetAgentModel is which model answers as them, changed on somebody who already exists. Refused
// for a model the app does not serve, the way hiring is — an agent pointed at nothing is a name
// that answers nothing, and it would only be found out at the next thing said to them.
func (c *Context) SetAgentModel(id, model string) (chat.Agent, error) {
	held, err := c.requireAgent(id)
	if err != nil {
		return chat.Agent{}, err
	}
	if chat.ProviderOf(model) == "" {
		return chat.Agent{}, apperr.Chatf("no such model: %s", model)
	}
	c.Chats.SetAgentModel(held.ID, model)
	again, _ := c.Chats.Agent(held.ID)
	return again, nil
}

// SetLead is who somebody reports to, or nobody. A loop is refused by walking upward from the
// proposed lead: meeting the agent on the way is the line closing on itself, and a chart asked
// who to escalate to would have no answer. Both ends have to be in this project — the chart is
// per-project the way the agents are.
func (c *Context) SetLead(agent string, lead *string) error {
	open, err := c.Workspace.RequireProject()
	if err != nil {
		return err
	}
	if c.Chats == nil {
		return apperr.Chatf("no such agent")
	}
	chart := map[string]chat.Placed{}
	for _, one := range c.Chats.Org(open.Path) {
		chart[one.ID] = one
	}
	held, found := chart[agent]
	if !found {
		return apperr.Chatf("no such agent")
	}
	if lead == nil {
		c.Chats.SetLead(agent, nil)
		return nil
	}
	if *lead == agent {
		return apperr.Chatf("%s cannot report to themselves", held.Name)
	}
	above, found := chart[*lead]
	if !found {
		return apperr.Chatf("no such agent")
	}
	for step := above.Lead; step != nil; {
		if *step == agent {
			return apperr.Chatf("that would make a loop: %s already reports to %s", above.Name, held.Name)
		}
		next, found := chart[*step]
		if !found {
			break
		}
		step = next.Lead
	}
	c.Chats.SetLead(agent, lead)
	return nil
}

// PlaceAgent is where they stand, after a drag.
func (c *Context) PlaceAgent(agent string, x, y float64) error {
	held, err := c.requireAgent(agent)
	if err != nil {
		return err
	}
	c.Chats.PlaceAgent(held.ID, x, y)
	return nil
}

func (c *Context) requireAgent(id string) (chat.Agent, error) {
	if c.Chats == nil {
		return chat.Agent{}, apperr.Chatf("no such agent")
	}
	held, found := c.Chats.Agent(id)
	if !found {
		return chat.Agent{}, apperr.Chatf("no such agent")
	}
	return held, nil
}
